python/cudf/cudf/tests/test_groupby.py

```diff
@@ -376,6 +377,142 @@ def emulate(df):
     assert_groupby_results_equal(expect, got)
 
 
+@pytest.fixture(scope="module")
+def groupby_jit_data():
+    np.random.seed(0)
+    df = DataFrame()
+    nelem = 20
+    df["key1"] = np.random.randint(0, 3, nelem)
+    df["key2"] = np.random.randint(0, 2, nelem)
+    df["val1"] = np.random.random(nelem)
+    df["val2"] = np.random.random(nelem)
+    return df
+
+
+def run_groupby_apply_jit_test(data, func, keys, *args):
+    expect_groupby_obj = data.to_pandas().groupby(keys, as_index=False)
+    got_groupby_obj = data.groupby(keys)
+
+    # compare cuDF jit to pandas
+    cudf_jit_result = got_groupby_obj.apply(func, *args, engine="jit")
+    pandas_result = expect_groupby_obj.apply(func, *args)
+    assert_groupby_results_equal(cudf_jit_result, pandas_result)
+
+
+@pytest.mark.parametrize("dtype", SUPPORTED_GROUPBY_NUMPY_TYPES)
+@pytest.mark.parametrize(
+    "func", ["min", "max", "sum", "mean", "var", "std", "idxmin", "idxmax"]
+)
+def test_groupby_apply_jit_reductions(func, groupby_jit_data, dtype):
+    # ideally we'd just have:
+    # lambda group: getattr(group, func)()
+    # but the current kernel caching mechanism relies on pickle which
+    # does not play nice with local functions. What's below uses
+    # exec as a workaround to write the test functions dynamically
+
+    funcstr = f"""
+def func(df):
+    return df['val1'].{func}()
+    """
+    lcl = {}
+    exec(funcstr, lcl)
+    func = lcl["func"]
+
+    groupby_jit_data["val1"] = groupby_jit_data["val1"].astype(dtype)
+    groupby_jit_data["val2"] = groupby_jit_data["val2"].astype(dtype)
+
+    run_groupby_apply_jit_test(groupby_jit_data, func, ["key1"])
+
+
+@pytest.mark.parametrize("dtype", ["float64"])
+@pytest.mark.parametrize("func", ["min", "max", "sum", "mean", "var", "std"])
+@pytest.mark.parametrize("special_val", [np.nan, np.inf, -np.inf])
+def test_groupby_apply_jit_reductions_special_vals(
+    func, groupby_jit_data, dtype, special_val
+):
+    # dynamically generate to avoid pickling error.
+    # see test_groupby_apply_jit_reductions for details.
+    funcstr = f"""
+def func(df):
+    return df['val1'].{func}()
+    """
```

Comment:
I'd like to avoid the manual wrapping that ends up going to the left of the test function's primary indention level. It reduces readability, especially if syntax highlighting is not available. I would write this as:

`import textwrap`, then:
```suggestion
    funcstr = textwrap.dedent(
        f"""\
        def func(df):
            return df['val1'].{func}()
        """
    )
```

This pattern is [followed in much of cudf](https://github.com/search?q=repo%3Arapidsai%2Fcudf%20textwrap.dedent&type=code).

Reply:
TIL! updating this.